src/charm.py

```diff
@@ -1,37 +1,58 @@
 #!/usr/bin/env python3
 
-# Copyright 2020 Canonical Ltd.
+# Copyright 2022 Canonical Ltd.
 # Licensed under the GPLv3, see LICENCE file for details.
 
 import logging
 import os
 import yaml
+import typing
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
-from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from ops import model, charm, framework, main
 
 
 logger = logging.getLogger()
 
 
-class JenkinsAgentCharm(CharmBase):
-    _stored = StoredState()
+class JenkinsAgentCharStoredState(framework.StoredState):
+    """Defines valid attributes of the stored state for the Jenkins Agent."""
+
+    relation_configured: bool | None
+    jenkins_url: str | None
+    agents: list[str] | None
+    agent_tokens: list[str] | None
+
+
+class JenkinsAgentEnvConfig(typing.TypedDict):
+    """The envornment configuration for the jenkins agent."""
+
+    JENKINS_AGENTS: str
+    JENKINS_TOKENS: str
+    JENKINS_URL: str
+
+
+class JenkinsAgentCharm(charm.CharmBase):
+    _stored = JenkinsAgentCharStoredState()
+    service_name = "jenkins-agent"
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.start, self._on_config_changed)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.upgrade_charm, self._on_config_changed)
+        self.framework.observe(self.on.start, self.on_config_changed)
```

Comment:
Isn't it recommended to have your handlers declared as private i.e. starting with an underscore? https://juju.is/docs/sdk/styleguide#heading--event-handler-visibility

Reply:
Thank you, fixed.

> Looks really good, solid work! I left just small remarks.
> There is one thing that I wanted to discuss - I see you run `tox -e unit`, and that is good - I was using it as well when converting a charm. I had to transfer some files (charmcraft.yaml most notably) to be able to `charmcraft pack` the converted charm. That's when I noticed the presence of `./run_tests` - it is generated by default for new charms by `charmcraft init`. It runs flake8 and unittest with coverage report and doesn't seem to rely on tox in any way. I'm wondering what is the recommended way of testing charms, as https://juju.is/docs/sdk/styleguide#heading--testing still favors tox.

Yeh I noticed that too, in this repo there is also a makefile which seems to have some value. My preference is also tox rather than the ./run_tests command since there is a range of tests (unit and integration) which are easily managed by tox. The indico charm is also using tox, maybe @mthaddon  can enlighten us :)